File: java/testsuite/ApplicationCases/SceneEditorApplication/src/model/Scene.java
package model;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.common.linealAlgebra.Matrix4x4d;
import vsdk.toolkit.environment.geometry.element.Ray;
import vsdk.toolkit.environment.material.RendererConfiguration;
import vsdk.toolkit.gui.feedback.ProgressMonitorConsole;
import vsdk.toolkit.media.RGBImageUncompressed;
import vsdk.toolkit.media.RGBAImageUncompressed;
import vsdk.toolkit.io.image.ImagePersistence;
import vsdk.toolkit.environment.material.SimpleMaterial;
import vsdk.toolkit.environment.camera.Camera;
import vsdk.toolkit.environment.camera.CameraSnapshot;
import vsdk.toolkit.environment.background.Background;
import vsdk.toolkit.environment.background.SimpleBackground;
import vsdk.toolkit.environment.background.CubemapBackground;
import vsdk.toolkit.environment.background.FixedBackground;
import vsdk.toolkit.environment.geometry.Geometry;
import vsdk.toolkit.environment.geometry.element.RayHit;
import vsdk.toolkit.environment.light.Light;
import vsdk.toolkit.environment.scene.SimpleBody;
import vsdk.toolkit.environment.scene.SimpleBodyGroup;
import vsdk.toolkit.environment.scene.SimpleScene;
import vsdk.toolkit.environment.scene.SimpleSceneSnapshot;
import vsdk.toolkit.render.jogl.Jogl4LightRenderer;
import vsdk.toolkit.render.raytracing.SimpleRaytracer;

// Application classes
import vsdk.toolkit.fixtures.Jogl2SimpleCorridorSample;

public class Scene
{
    public SimpleScene scene;

    //- 1. Camera ----------------------------------------------------------
    public Camera camera;
    public Camera activeCamera;

    //- 3. Background ------------------------------------------------------
    public SimpleBackground simpleBackground;
    public CubemapBackground cubemapBackground;
    public FixedBackground fixedBackground;
    public int selectedBackground;

    //- 4. Objects ---------------------------------------------------------
    public Jogl2SimpleCorridorSample corridor;
    public boolean showCorridor;
    public ArrayList<SimpleBodyGroup> debugThingGroups;

    public SelectionSet selectedThings;
    public SelectionSet selectedLights;
    public SelectionSet selectedDebugThingGroups;

    // Others
    public RendererConfiguration qualityTemplate;
    private int acumObject = 1;

    public Scene()
    {
        scene = new SimpleScene();

        //-----------------------------------------------------------------
        debugThingGroups = new ArrayList<SimpleBodyGroup>();

        Matrix4x4d R = new Matrix4x4d();
        camera = new Camera();

        R = R.eulerAnglesRotation(Math.toRadians(45), Math.toRadians(-35), 0);
        camera.setPosition(new Vector3Dd(-5, -5, 5));
        camera.setRotation(R);

        activeCamera = camera;
        selectedThings = new SelectionSet(scene.getSimpleBodies());
        selectedLights = new SelectionSet(scene.getLights());
        selectedDebugThingGroups = new SelectionSet(debugThingGroups);

        //-----------------------------------------------------------------
        simpleBackground = new SimpleBackground();
        simpleBackground.setColor(0.49, 0.49, 0.49);

        cubemapBackground = null;
        fixedBackground = null;

        selectedBackground = 0;

        //-----------------------------------------------------------------
        corridor = new Jogl2SimpleCorridorSample();
        showCorridor = false;

        qualityTemplate = new RendererConfiguration();
        qualityTemplate.setSurfaces(true);
        qualityTemplate.setWires(false);
    }

    public boolean
    buildCubemap()
    {
        RGBAImageUncompressed front, right, back, left, down, up;

        try {
            System.out.print("Loading background: 1");
            front = ImagePersistence.importRGBA(
                        new File("../../../../etc/cubemaps/dorise1/entorno0.jpg"));
            System.out.print("2");
            right = ImagePersistence.importRGBA(
                        new File("../../../../etc/cubemaps/dorise1/entorno1.jpg"));
            System.out.print("3");
            back = ImagePersistence.importRGBA(
                        new File("../../../../etc/cubemaps/dorise1/entorno2.jpg"));
            System.out.print("4");
            left = ImagePersistence.importRGBA(
                        new File("../../../../etc/cubemaps/dorise1/entorno3.jpg"));
            System.out.print("5");
            down = ImagePersistence.importRGBA(
                        new File("../../../../etc/cubemaps/dorise1/entorno4.jpg"));
            System.out.print("6");
            up = ImagePersistence.importRGBA(
                        new File("../../../../etc/cubemaps/dorise1/entorno5.jpg"));
            System.out.println(" OK!");

            cubemapBackground = 
                new CubemapBackground(camera, 
                                      front, right, back, left, down, up);
        }
        catch (Exception e) {
            System.err.println(e);
            return false;
        }
        return true;
    }

    public boolean
    buildFixedmap()
    {
        RGBAImageUncompressed img;

        try {
            System.out.print("Loading background: ");
            img = ImagePersistence.importRGBA(
                        new File("../../../../etc/cubemaps/dorise1/entorno0.jpg"));
            System.out.println("OK!");

            fixedBackground = new FixedBackground(camera, img);
        }
        catch (Exception e) {
            System.err.println(e);
            return false;
        }
        return true;
    }

    public SimpleMaterial defaultMaterial()
    {
        SimpleMaterial m = new SimpleMaterial();

/*
        m = m.withAmbient(new ColorRgb(0.2, 0.2, 0.2));
        m = m.withDiffuse(new ColorRgb(0.5, 0.9, 0.5));
        m = m.withSpecular(new ColorRgb(1, 1, 1));
        m = m.withDoubleSided(false);
        m = m.withPhongExponent(100.0);
*/

        m = m.withAmbient(new ColorRgb(0, 0, 0));
        m = m.withDiffuse(new ColorRgb(1, 1, 1));
        m = m.withSpecular(new ColorRgb(1, 1, 1));
        m = m.withDoubleSided(false);
        m = m.withPhongExponent(40.0);


        return m;
    }

    public SimpleBody addThing(Geometry g)
    {
        SimpleBody thing;

        thing = new SimpleBody();
        thing.setGeometry(g);
        thing.setPosition(new Vector3Dd());
        thing.setRotation(new Matrix4x4d());
        thing.setRotationInverse(new Matrix4x4d());
        thing.setMaterial(defaultMaterial());
        thing.setName("Geometric object " + acumObject);
        scene.getSimpleBodies().add(thing);

        acumObject++;
        selectedThings.sync();
        return thing;
    }

    public boolean doIntersectionFirstHit(Ray r, RayHit info)
    {
        int i;
        double nearestDistance = Float.MAX_VALUE;
        List<SimpleBody> things = scene.getSimpleBodies();
        boolean intersected = false;
        SimpleBody gi;
        RayHit ii = new RayHit();
        for ( i = 0; i < things.size(); i++ ) {
            gi = things.get(i);
            RayHit hit = new RayHit();
            if ( gi.doIntersectionFirstHit(r, hit) && hit.ray().getT() < nearestDistance ) {
                ii.clone(hit);
                nearestDistance = hit.ray().getT();
                r = hit.ray();
                intersected = true;
            }
        }
        if ( intersected ) {
            r = r.withT(nearestDistance);
            info.clone(ii);
            return true;
        }
        return false;
    }

    /**
    Selects the nearest thing (body or light) under a pixel of the active
    camera viewport. Without `composite` the previous selection is discarded,
    with it the picked thing changes its selection state.
    Bodies are picked with their geometry and lights with a sphere of the size
    of their gizmo (see `LightPicker`).
    @param x pixel column in the viewport
    @param y pixel row in the viewport
    @param composite true to modify the current selection
    @param ro ray to be replaced
    @return the ray fired through the pixel
    */
    public Ray selectObjectWithMouse(int x, int y, boolean composite, Ray ro)
    {
        Ray r;
        SimpleBody gi;

        activeCamera.updateVectors();
        r = activeCamera.generateRay(x, y);

        Ray selectedRay = Ray.copyOf(r);

        double nearestDistance = Float.MAX_VALUE;
        int nearestBody = -1;
        int nearestLight = -1;

        int i;

        selectedThings.sync();
        selectedLights.sync();

        List<SimpleBody> things = scene.getSimpleBodies();
        for ( i = 0; i < things.size(); i++ ) {
            gi = things.get(i);
            Ray hit = gi.doIntersectionFirstHit(r);
            if ( hit != null && hit.getT() < nearestDistance ) {
                nearestDistance = hit.getT();
                nearestBody = i;
            }
        }

        List<Light> lights = scene.getLights();
        for ( i = 0; i < lights.size(); i++ ) {
            double t = LightPicker.pick(r, activeCamera, lights.get(i),
                Jogl4LightRenderer.getScale());
            if ( t >= 0 && t < nearestDistance ) {
                nearestDistance = t;
                nearestBody = -1;
                nearestLight = i;
            }
        }

        if ( !composite ) {
            selectedThings.unselectAll();
            selectedLights.unselectAll();
            selectedThings.select(nearestBody);
            selectedLights.select(nearestLight);
        }
        else {
            selectedThings.change(nearestBody);
            selectedLights.change(nearestLight);
        }
        return selectedRay;
    }

    /**
    Selects the next background, cycling over the three available ones.
    */
    public void rotateBackground()
    {
        selectedBackground++;
        if ( selectedBackground > 2 ) {
            selectedBackground = 0;
        }
    }

    public void activateSelectedBackground()
    {
        Background currentBackground;

        currentBackground = simpleBackground;
        switch ( selectedBackground ) {
          case 2:
            if ( cubemapBackground == null ) {
                buildCubemap();
            }
            if ( cubemapBackground != null ) {
                cubemapBackground.setCamera(activeCamera);
                currentBackground = cubemapBackground;
            }
            break;
          case 1:
            if ( fixedBackground == null ) {
                buildFixedmap();
            }
            if ( fixedBackground != null ) {
                currentBackground = fixedBackground;
            }
            break;
        }

        ArrayList<Background> list;
        list = scene.getBackgrounds();
        if ( list.size() < 1 ) {
            list.add(currentBackground);
        }
        else {
            list.remove(0);
            list.add(0, currentBackground);
        }
        scene.setActiveBackgroundIndex(0);
    }

    public void print()
    {
        List<SimpleBody> things = scene.getSimpleBodies();
        int i;

        System.out.println("= SCENE REPORT ============================================================");
        System.out.println("Current camera:\n" +  activeCamera);
        System.out.println("Things in scene: " + things.size());
        for ( i = 0; i < things.size(); i++ ) {
            System.out.println("  - Thing[" + i + "]: " + things.get(i).getGeometry().getClass().getName());
        }
        System.out.println("= END OF REPORT ===========================================================");
    }

    public void raytrace(RGBImageUncompressed out_Viewport)
    {
        int originalWidth;
        int originalHeight;

        originalWidth = (int)activeCamera.getViewportXSize();
        originalHeight = (int)activeCamera.getViewportYSize();
        CameraSnapshot cameraSnapshot = activeCamera.exportToCameraSnapshot(
            out_Viewport.getXSize(), out_Viewport.getYSize());

        //-----------------------------------------------------------------
        ProgressMonitorConsole reporter = new ProgressMonitorConsole();        
        SimpleRaytracer visualizationEngine;

        Background activeBackground;
        switch ( selectedBackground ) {
          case 2:
            if ( cubemapBackground == null ) {
                buildCubemap();
            }
            if ( cubemapBackground != null ) {
                activeBackground = cubemapBackground;
            }
            else {
                activeBackground = simpleBackground;
            }
            break;
          case 1:
            if ( fixedBackground == null ) {
                buildFixedmap();
            }
            if ( fixedBackground != null ) {
                activeBackground = fixedBackground;
            }
            else {
                activeBackground = simpleBackground;
            }
            break;
          case 0: default:
            activeBackground = simpleBackground;
            break;
        }

        visualizationEngine = new SimpleRaytracer();
        SimpleSceneSnapshot sceneSnapshot =
            scene.exportToSimpleSceneSnapshot(cameraSnapshot, activeBackground);
        long initialTime = System.currentTimeMillis();
        visualizationEngine.execute(out_Viewport, qualityTemplate,
                                    sceneSnapshot, reporter, null);
        long finalTime = System.currentTimeMillis();
        System.out.println("Image generated in " + (finalTime-initialTime) + " miliseconds.");

        File fd = new File("./output.jpg");

        System.out.print("Exporting result image to file: ");
        if ( !ImagePersistence.exportJPG(fd, out_Viewport) )
        {
            System.err.println("Error grabando la imagen!!");
            System.exit(1);
        }
        System.out.println(" OK!");
        System.out.println("An image has been created in the file output.jpg");

        //-----------------------------------------------------------------
        activeCamera.updateViewportResize(originalWidth, originalHeight);
    }

}

File: cpp/testsuite/ApplicationCases/SceneEditorApplication/src/model/Scene.cpp
#include <cfloat>
#include <cmath>
#include <typeinfo>

#include "java/io/File.h"
#include "java/lang/System.h"
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/common/color/ColorRgb.h"
#include "vsdk/toolkit/common/linealAlgebra/Matrix4x4d.h"
#include "vsdk/toolkit/common/linealAlgebra/Vector3Dd.h"
#include "vsdk/toolkit/environment/background/CubemapBackground.h"
#include "vsdk/toolkit/environment/background/FixedBackground.h"
#include "vsdk/toolkit/environment/background/SimpleBackground.h"
#include "vsdk/toolkit/environment/camera/Camera.h"
#include "vsdk/toolkit/environment/camera/CameraSnapshot.h"
#include "vsdk/toolkit/environment/geometry/Geometry.h"
#include "vsdk/toolkit/environment/geometry/element/Ray.h"
#include "vsdk/toolkit/environment/geometry/element/RayHit.h"
#include "vsdk/toolkit/environment/light/Light.h"
#include "vsdk/toolkit/environment/material/RendererConfiguration.h"
#include "vsdk/toolkit/environment/material/SimpleMaterial.h"
#include "vsdk/toolkit/environment/scene/SimpleBody.h"
#include "vsdk/toolkit/environment/scene/SimpleBodyGroup.h"
#include "vsdk/toolkit/environment/scene/SimpleScene.h"
#include "vsdk/toolkit/environment/scene/SimpleSceneSnapshot.h"
#include "vsdk/toolkit/io/image/ImagePersistence.h"
#include "vsdk/toolkit/media/RGBAImageUncompressed.h"
#include "vsdk/toolkit/media/RGBImageUncompressed.h"
#include "vsdk/toolkit/media/ZBuffer.h"
#include "vsdk/toolkit/render/raytracing/DepthBufferMode.h"
#include "vsdk/toolkit/render/raytracing/ParallelRaytracer.h"

#include "model/Scene.h"
#include "model/selection/SelectionSet.h"

namespace {
double toRadians(double degrees)
{
    return degrees * M_PI / 180.0;
}

/// Distance of a hit along its ray (the Java version reads the ray stored
/// in the hit)
double hitDistance(const RayHit& hit)
{
    if ( hit.hasHitDistance() ) {
        return hit.getHitDistance();
    }
    if ( hit.getRay() != nullptr ) {
        return hit.getRay()->getT();
    }
    return DBL_MAX;
}

RGBAImageUncompressed* importCubemapFace(const char* fileName)
{
    return ImagePersistence::importRGBA(java::File(fileName));
}
}

ParallelRaytracer* Scene::getRaytracer()
{
    static ParallelRaytracer raytracer;
    return &raytracer;
}

Scene::Scene() : acumObject(1), lightGizmoScale(1.0)
{
    scene = new SimpleScene();

    //-----------------------------------------------------------------
    Matrix4x4d R;
    camera = new Camera();

    R = R.eulerAnglesRotation(toRadians(45), toRadians(-35), 0);
    camera->setPosition(Vector3Dd(-5, -5, 5));
    camera->setRotation(R);

    activeCamera = camera;
    selectedThings = new SelectionSet(&scene->getSimpleBodies());
    selectedLights = new SelectionSet(&scene->getLights());
    selectedDebugThingGroups = new SelectionSet(&debugThingGroups);

    //-----------------------------------------------------------------
    simpleBackground = new SimpleBackground();
    simpleBackground->setColor(0.49, 0.49, 0.49);

    cubemapBackground = nullptr;
    fixedBackground = nullptr;

    selectedBackground = 0;

    //-----------------------------------------------------------------
    showCorridor = false;

    qualityTemplate = new RendererConfiguration();
    qualityTemplate->setSurfaces(true);
    qualityTemplate->setWires(false);
}

Scene::~Scene()
{
    long i;

    delete selectedThings;
    delete selectedLights;
    delete selectedDebugThingGroups;

    // Backgrounds are owned by this scene, not by the vitral scene
    scene->getBackgrounds().clear();
    delete scene;
    delete camera;
    delete simpleBackground;
    delete cubemapBackground;
    delete fixedBackground;
    // Debug groups own their bodies (see `ProjectedViewsDebugger`)
    for ( i = 0; i < debugThingGroups.size(); i++ ) {
        java::ArrayList<SimpleBody*>& bodies =
            debugThingGroups.get(i)->getBodies();
        long j;
        for ( j = 0; j < bodies.size(); j++ ) {
            delete bodies.get(j);
        }
        delete debugThingGroups.get(i);
    }
    delete qualityTemplate;
}

double Scene::getLightGizmoScale() const
{
    return lightGizmoScale;
}

void Scene::setLightGizmoScale(double lightGizmoScale)
{
    this->lightGizmoScale = lightGizmoScale;
}

bool Scene::buildCubemap()
{
    RGBAImageUncompressed* front;
    RGBAImageUncompressed* right;
    RGBAImageUncompressed* back;
    RGBAImageUncompressed* left;
    RGBAImageUncompressed* down;
    RGBAImageUncompressed* up;

    java::System::out.print("Loading background: 1");
    front = importCubemapFace("../../../../etc/cubemaps/dorise1/entorno0.jpg");
    java::System::out.print("2");
    right = importCubemapFace("../../../../etc/cubemaps/dorise1/entorno1.jpg");
    java::System::out.print("3");
    back = importCubemapFace("../../../../etc/cubemaps/dorise1/entorno2.jpg");
    java::System::out.print("4");
    left = importCubemapFace("../../../../etc/cubemaps/dorise1/entorno3.jpg");
    java::System::out.print("5");
    down = importCubemapFace("../../../../etc/cubemaps/dorise1/entorno4.jpg");
    java::System::out.print("6");
    up = importCubemapFace("../../../../etc/cubemaps/dorise1/entorno5.jpg");

    if ( front == nullptr || right == nullptr || back == nullptr ||
         left == nullptr || down == nullptr || up == nullptr ) {
        java::System::err.println(" Error loading cubemap images");
        delete front;
        delete right;
        delete back;
        delete left;
        delete down;
        delete up;
        return false;
    }
    java::System::out.println(" OK!");

    // C++ port note: the images are referenced by the background, and kept
    // for the rest of the program (as done by the Java garbage collector)
    cubemapBackground =
        new CubemapBackground(camera, front, right, back, left, down, up);
    return true;
}

bool Scene::buildFixedmap()
{
    RGBAImageUncompressed* img;

    java::System::out.print("Loading background: ");
    img = importCubemapFace("../../../../etc/cubemaps/dorise1/entorno0.jpg");
    if ( img == nullptr ) {
        java::System::err.println("Error loading background image");
        return false;
    }
    java::System::out.println("OK!");

    fixedBackground = new FixedBackground(camera, img);
    return true;
}

SimpleMaterial Scene::defaultMaterial()
{
    SimpleMaterial m;

/*
    m = m.withAmbient(ColorRgb(0.2, 0.2, 0.2));
    m = m.withDiffuse(ColorRgb(0.5, 0.9, 0.5));
    m = m.withSpecular(ColorRgb(1, 1, 1));
    m = m.withDoubleSided(false);
    m = m.withPhongExponent(100.0);
*/

    m = m.withAmbient(ColorRgb(0, 0, 0));
    m = m.withDiffuse(ColorRgb(1, 1, 1));
    m = m.withSpecular(ColorRgb(1, 1, 1));
    m = m.withDoubleSided(false);
    m = m.withPhongExponent(40.0);

    return m;
}

SimpleBody* Scene::addThing(Geometry* g)
{
    SimpleBody* thing;

    thing = new SimpleBody();
    thing->setGeometry(g);
    thing->setPosition(Vector3Dd());
    thing->setRotation(Matrix4x4d());
    thing->setRotationInverse(Matrix4x4d());
    thing->setMaterial(new SimpleMaterial(defaultMaterial()));
    thing->setName(java::String("Geometric object ") +
                   java::String::valueOf(acumObject));
    scene->getSimpleBodies().add(thing);

    acumObject++;
    selectedThings->sync();
    return thing;
}

bool Scene::doIntersectionFirstHit(const Ray& inRay, RayHit* info)
{
    int i;
    double nearestDistance = FLT_MAX;
    java::ArrayList<SimpleBody*>& things = scene->getSimpleBodies();
    bool intersected = false;
    SimpleBody* gi;
    RayHit ii;
    Ray r(inRay);

    for ( i = 0; i < things.size(); i++ ) {
        gi = things.get(i);
        RayHit hit;
        hit.setStoreRay(true);
        if ( gi->doIntersectionFirstHit(r, &hit) ) {
            double t = hitDistance(hit);
            if ( t < nearestDistance ) {
                ii.clone(hit);
                nearestDistance = t;
                r = r.withT(t);
                intersected = true;
            }
        }
    }
    if ( intersected ) {
        info->clone(ii);
        return true;
    }
    return false;
}

void Scene::rotateBackground()
{
    selectedBackground++;
    if ( selectedBackground > 2 ) {
        selectedBackground = 0;
    }
}

void Scene::activateSelectedBackground()
{
    Background* currentBackground;

    currentBackground = simpleBackground;
    switch ( selectedBackground ) {
      case 2:
        if ( cubemapBackground == nullptr ) {
            buildCubemap();
        }
        if ( cubemapBackground != nullptr ) {
            cubemapBackground->setCamera(activeCamera);
            currentBackground = cubemapBackground;
        }
        break;
      case 1:
        if ( fixedBackground == nullptr ) {
            buildFixedmap();
        }
        if ( fixedBackground != nullptr ) {
            currentBackground = fixedBackground;
        }
        break;
    }

    java::ArrayList<Background*>& list = scene->getBackgrounds();
    if ( list.size() < 1 ) {
        list.add(currentBackground);
    }
    else {
        list.remove((long)0);
        list.add((long)0, currentBackground);
    }
    scene->setActiveBackgroundIndex(0);
}

void Scene::print()
{
    java::ArrayList<SimpleBody*>& things = scene->getSimpleBodies();
    int i;

    java::System::out.println("= SCENE REPORT ============================================================");
    java::System::out.println((java::String("Current camera:\n") +
                               activeCamera->toString()).c_str());
    java::System::out.println((java::String("Things in scene: ") +
        java::String::valueOf((long)things.size())).c_str());
    for ( i = 0; i < things.size(); i++ ) {
        Geometry* g = things.get(i)->getGeometry();
        java::String msg = java::String("  - Thing[") +
            java::String::valueOf(i) + "]: " +
            (g != nullptr ? typeid(*g).name() : "null");
        java::System::out.println(msg.c_str());
    }
    java::System::out.println("= END OF REPORT ===========================================================");
}

void Scene::raytrace(RGBImageUncompressed* outViewport)
{
    raytrace(outViewport, nullptr, true);
}

void Scene::raytraceViewport(RGBImageUncompressed* outViewport)
{
    raytrace(outViewport, nullptr, false);
}

void Scene::raytraceViewport(RGBImageUncompressed* outViewport,
                             ZBuffer* outDepth)
{
    raytrace(outViewport, outDepth, false);
}

void Scene::raytrace(RGBImageUncompressed* outViewport, ZBuffer* outDepth,
                     bool interactiveReport)
{
    int originalWidth;
    int originalHeight;

    originalWidth = (int)activeCamera->getViewportXSize();
    originalHeight = (int)activeCamera->getViewportYSize();
    CameraSnapshot* cameraSnapshot = activeCamera->exportToCameraSnapshot(
        outViewport->getXSize(), outViewport->getYSize());

    //-----------------------------------------------------------------

    Background* activeBackground;
    switch ( selectedBackground ) {
      case 2:
        if ( cubemapBackground == nullptr ) {
            buildCubemap();
        }
        if ( cubemapBackground != nullptr ) {
            activeBackground = cubemapBackground;
        }
        else {
            activeBackground = simpleBackground;
        }
        break;
      case 1:
        if ( fixedBackground == nullptr ) {
            buildFixedmap();
        }
        if ( fixedBackground != nullptr ) {
            activeBackground = fixedBackground;
        }
        else {
            activeBackground = simpleBackground;
        }
        break;
      case 0: default:
        activeBackground = simpleBackground;
        break;
    }

    // The snapshot owns the camera snapshot
    SimpleSceneSnapshot* sceneSnapshot =
        scene->exportToSimpleSceneSnapshot(cameraSnapshot, activeBackground);
    long long initialTime = java::System::currentTimeMillis();
    getRaytracer()->execute(outViewport, outDepth,
        outDepth != nullptr ? DepthBufferMode::OPENGL_DEPTH :
                              DepthBufferMode::NONE,
        qualityTemplate, sceneSnapshot, interactiveReport);
    long long finalTime = java::System::currentTimeMillis();
    delete sceneSnapshot;

    if ( interactiveReport ) {
        java::String msg = java::String("Image generated in ") +
            java::String::valueOf(finalTime-initialTime) + " miliseconds.";
        java::System::out.println(msg.c_str());

        java::File fd("./output.jpg");

        java::System::out.print("Exporting result image to file: ");
        if ( !ImagePersistence::exportJPEG(fd, outViewport) ) {
            java::System::err.println("Error grabando la imagen!!");
            java::System::exit(1);
        }
        java::System::out.println(" OK!");
        java::System::out.println("An image has been created in the file output.jpg");
    }

    //-----------------------------------------------------------------
    activeCamera->updateViewportResize(originalWidth, originalHeight);
}
